pkg/sdkserver/datasets.go

```diff
@@ -0,0 +1,332 @@
+package sdkserver
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	gcontext "github.com/gptscript-ai/gptscript/pkg/context"
+	"github.com/gptscript-ai/gptscript/pkg/gptscript"
+	"github.com/gptscript-ai/gptscript/pkg/loader"
+)
+
+type datasetRequest struct {
+	Input           string `json:"input"`
+	Workspace       string `json:"workspace"`
+	DatasetToolRepo string `json:"datasetToolRepo"`
+}
+
+func (r datasetRequest) validate(requireInput bool) error {
+	if r.Workspace == "" {
+		return fmt.Errorf("workspace is required")
+	} else if requireInput && r.Input == "" {
+		return fmt.Errorf("input is required")
+	}
+	return nil
+}
+
+func (r datasetRequest) opts(o gptscript.Options) gptscript.Options {
+	opts := gptscript.Options{
+		Cache:     o.Cache,
+		Monitor:   o.Monitor,
+		Runner:    o.Runner,
+		Workspace: r.Workspace,
+	}
+	return opts
+}
+
+func (r datasetRequest) getToolRepo() string {
+	if r.DatasetToolRepo != "" {
+		return r.DatasetToolRepo
+	}
+	return "github.com/gptscript-ai/datasets"
+}
+
+func (s *server) listDatasets(w http.ResponseWriter, r *http.Request) {
+	logger := gcontext.GetLogger(r.Context())
+
+	var req datasetRequest
+	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
+		writeError(logger, w, http.StatusBadRequest, fmt.Errorf("failed to decode request body: %w", err))
+		return
+	}
+
+	if err := req.validate(false); err != nil {
+		writeError(logger, w, http.StatusBadRequest, err)
+		return
+	}
+
+	g, err := gptscript.New(r.Context(), req.opts(s.gptscriptOpts))
+	if err != nil {
+		writeError(logger, w, http.StatusInternalServerError, fmt.Errorf("failed to initialize gptscript: %w", err))
+		return
+	}
+
+	prg, err := loader.Program(r.Context(), "List Datasets from "+req.getToolRepo(), "", loader.Options{
+		Cache: g.Cache,
+	})
+
+	if err != nil {
+		writeError(logger, w, http.StatusInternalServerError, fmt.Errorf("failed to load program: %w", err))
+		return
+	}
+
+	result, err := g.Run(r.Context(), prg, s.gptscriptOpts.Env, req.Input)
+	if err != nil {
+		writeError(logger, w, http.StatusInternalServerError, fmt.Errorf("failed to run program: %w", err))
+		return
+	}
+
+	writeResponse(logger, w, map[string]any{"stdout": result})
+}
+
+type createDatasetArgs struct {
+	Name        string `json:"dataset_name"`
+	Description string `json:"dataset_description"`
```

Comment:
nit: you have camelCase and snake_case in your `json` tags. Can this be standardized?

Reply:
fixed

Reply:
This went the opposite way I expected. Is there reason why you are doing snake_case? I think most of our stuff uses camelCase.

Reply:
Lol. I honestly have no idea. I'll go change it all to camelCase.

Reply:
Fixed. Sorry for the delay. Got stuck working on a test for another repo